kanisterio/kanister
# Add example/blueprint for etcd backup

## Change Overview

This PR add steps of how do we backup the etcd that is deployed as part of kubernetes control plane.

## Pull request type

Please check the type of change your PR introduces:

## Issues

- #XXX

## Test Plan
- SSH into the host where ETCD is running.
- Static pods are usually at /etc/kubernetes/manifests
- Install etcdctl (http://launchpadlibrarian.net/361589174/etcd-client_3.2.17+dfsg-1_amd64.deb)
- Create a dummy namespace and dummy deployment in that namespace
- Backup the etcd instance using the backup action of the provided blueprint, which eventually does below
```
    ETCDCTL_API=3 etcdctl --endpoints=https://127.0.0.1:2379 --cacert=/etc/kubernetes/pki/etcd/ca.crt  --cert=/etc/kubernetes/pki/etcd/server.crt  --key=/etc/kubernetes/pki/etcd/server.key snapshot save /var/lib/etcd/etcd-backup.db

    ETCDCTL_API=3 etcdctl --endpoints=https://127.0.0.1:2379 --cacert=/etc/kubernetes/pki/etcd/ca.crt  --cert=/etc/kubernetes/pki/etcd/server.crt  --key=/etc/kubernetes/pki/etcd/server.key snapshot status /var/lib/etcd/etcd-backup.db
```
- delete that namespace we created
- restore the etcd following the steps mentioned here and docs
```
ETCDCTL_API=3 etcdctl --endpoints=https://127.0.0.1:2379 --cacert=/etc/kubernetes/pki/etcd/ca.crt --cert=/etc/kubernetes/pki/etcd/server.crt  --key=/etc/kubernetes/pki/etcd/server.key --data-dir="/var/lib/etcd-from-backup" --initial-cluster="ubuntu-s-4vcpu-8gb-blr1-01-master-1=https://127.0.0.1:2380" --name="ubuntu-s-4vcpu-8gb-blr1-01-master-1"  --initial-advertise-peer-urls="https://127.0.0.1:2380" --initial-cluster-token="etcd-cluster-1"  snapshot restore /tmp/etcd-backup.db
```
- And follow readme to make changes in the etcd static pod
- Once etcd pod is running with new configuration, make sure the namespace is restored back.
